base-kv/base-kv-store-balance-controller/src/main/java/com/baidu/bifromq/basekv/balance/KVRangeBalanceController.java

```diff
@@ -155,26 +175,25 @@ private void scheduleNow() {
                     String cmdName = commandToRun.getClass().getSimpleName();
                     Sample start = Timer.start();
                     CommandType commandType = commandToRun.type();
-                    commandRunner.run(commandToRun)
+                    runCommand(commandToRun)
                         .whenCompleteAsync((r, e) -> {
+                            scheduling.set(false);
                             CommandMetrics metrics = metricsManager.getCommandMetrics(balancerName, cmdName);
-                            if (r == CommandRunner.Result.Succeed) {
+                            if (Boolean.TRUE.equals(r)) {
                                 metrics.cmdSucceedCounter.increment();
                                 start.stop(metrics.cmdRunTimer);
                                 // Always schedule later after recovery command
                                 if (commandType == CommandType.RECOVERY) {
-                                    scheduling.set(false);
-                                    scheduleLater();
+                                    scheduleLater(randomDelay(), TimeUnit.MILLISECONDS);
                                 } else {
-                                    scheduleNow();
+                                    scheduleLater(1, TimeUnit.SECONDS);
```

Comment:
why treat recovery as a special case?

Reply:
The "RecoverRequest" does not have a "version" field, which could lead to it being regenerated and executed redundantly.